### ISSUE TITLE
Protocol: generation as a position frame — greedy decode addressed by `pos`

### ISSUE THREAD

**Issue by can-goodfire** [goodfire-ai/causalab#25]
## Context

The protocol refactor (#20) is prefill-only by declaration:
`docs/intervention_protocol.md:371` — *"**v1 scope**: prefill-only. No generation, no
decode-step writes, one neural model per document."* Nothing in the document vocabulary
decodes, which blocks the first thing an exploration loop does: greedy-decode a prompt batch
under an intervention, read the continuation, and check whether an edited token flips it.
The retired stack had this as a first-class surface
(`causalab/neural/dataset.py::run_intervened_generation` at base `a50637c`).

This issue adds greedy decode **without** relaxing prefill-only for interventions, and
without a new section: generation becomes a **position frame**, so reads address it and
metrics compose with it (metrics bind to reads — §2.10 — which is the whole point).

## The decision (settled 2026-08-21)

**Generation extends the position frame; `pos` addresses the continuation.**

A read stays a value at an address. What is new is *which frame the address is resolved
against*: positions are never integers in the document — they are resolved by the backend
against a `PositionFrame` (§2.3, §8) — and a decode simply produces a second frame. So the
position spec gains a **frame selector**, `generated`, orthogonal to the existing anchor set.

Rejected alternatives, briefly, so the choice is legible:

- **A new `generations` section** producing a tokens/text table: outputs generation, but
  nothing composes — metrics bind to reads, so scoring a continuation would need a parallel
  metric vocabulary.
- **An execution-mode field on `intervened_models`**: `original` is never declared (§2.9), so
  an un-intervened probe has nothing to hang it on (the `{"writes": []}` stand-in validates,
  but names "an intervened model with no interventions" and earns a second forward for
  identical activations — `plan.py:168` `_model_closure`); and a read of a generating model
  would *still* need a way to say "I mean the continuation", so it is strictly more
  vocabulary.
- **A `generate` field on the read** (sibling of `site`/`pos`): one field, but it splits the
  addressing story in two — `pos` for prompt positions, a second field for the continuation —
  and it cannot express *"the tokens where the model said X"*, which needs the full anchor
  grammar (`variable`, `span`, `all`, `index`) applied to the continuation.

## The spelling

The frame selector carries the decode budget; the anchor selects inside the frame. Idiomatic
use is the `positions` table, so one budget is declared once and shared by every read that
references it:

```json
"positions": {
  "continuation": {"generated": {"max_new_tokens": 32}, "all": true},
  "last_new":     {"generated": {"max_new_tokens": 32}, "index": -1},
  "said_answer":  {"generated": {"max_new_tokens": 32}, "variable": "expected"}
}
```

| form | resolves to |
|---|---|
| `{"generated": {"max_new_tokens": N}, "all": true}` | every generated token of the row |
| `+ {"index": -1}` | the **final** generated token (`0` = the first) |
| `+ {"span": [a, b]}` | a fixed window of the continuation |
| `+ {"variable": "x"}` | the tokens covering the row's value for `x` **inside the generated text** |

- `generated` is a frame selector, **not** an anchor: exactly one anchor
  (`index` \| `span` \| `variable` \| `all`) must accompany it, and the existing
  anchor-exclusivity check (`schema.py:857`) is unchanged.
- `max_new_tokens` is a required int ≥ 1, sweepable like any other field. Two positions on
  one `(model, input)` with different budgets are legal: the run decodes the max, each read
  windows its own; each budget is materialized per position in the canonical form, so digests
  separate them.
- `column` is refused with `generated` (a dataset-supplied *prompt* index means nothing in the
  continuation), as are `scope`/`relative_to` in v1 — `relative_to: {"variable": …}` inside
  the continuation is an obvious follow-up, deliberately not v1.

**Frame semantics**

- The continuation frame is the generated tokens of the row, ending at the **first EOS**
  (exclusive) or at `max_new_tokens`. Rows therefore have different widths — **ragged, which
  the stack already handles end to end** (`RaggedValue` in the executor, flat + `.widths` in
  `TensorFile.add`, both landed with #27). `index: -1` means the last *real* token, which is
  what "the final generated token" should mean. The decode may stop early once every row has
  hit EOS.
- **`lm_head` off-by-one, stated once so no document inherits it:** `lm_head` at generated
  position `j` is the distribution predicted *after* token `j`. The distribution that
  *produced* token `j` sits at position `j−1`, and for `j = 0` that is the last **prompt**
  position — addressable today as an ordinary `{"index": -1}`. No special case; it just has
  to be written down.
- **Writes may not carry `generated`** — a load error, and the rule that keeps prefill-only
  intact. Bonus simplification: `_provably_disjoint` (`validate.py:477`) becomes *less*
  conservative, since a prompt-frame and a continuation-frame position are provably disjoint.

**Variable matching inside the generation** (the interesting anchor):

- The value comes from the row exactly as on the prompt side — the sibling `<col>_variables`
  mapping or a same-named column (`encoding.py:123` `variable_value`).
- **First** occurrence, not "exactly once". The prompt-side uniqueness rule would turn a
  repetitive generation into a data-dependent crash.
- **Zero occurrences is a result, not an error**: that row contributes zero positions (width-0
  ragged), and metrics over it emit `null` plus a `matched` boolean column — so "the model
  never said it" is distinguishable from "it said it and scored 0". This is the one place
  where refusing loudly would be wrong: whether the model says the thing is the experiment.
- Char→token spans come from **incremental detokenization** (accumulating `decode(prefix)`
  lengths as the loop runs), not `offset_mapping` — sentencepiece merges make per-token
  offsets unavailable after the fact.

## Materialization: the document states the requirement, the backend picks the mechanism

The expensive object is the vocabulary projection, not the decode. At batch 32, `N = 16`, a
128k vocabulary in fp32: every step's distribution = **260 MB**, one step = **16 MB**, a site's
per-step activations (d=4096) = **8 MB**, the token ids = **2 KB**. So a document that wants the
final token's distribution must not cost the same as one that wants all of them.

**What this issue adds to the protocol is a derived requirement, not a strategy.** The planner
emits, per forward group: the **decode depth** (the max `max_new_tokens` over the group's
generate positions) and, per generate read, `{site, frame, needs_distribution}` where

```
needs_distribution(read) = read is saved
                        or any(m.of == read and m.domain == "distribution" for m in metrics)
```

with each metric kind declaring its domain — `ids` (the `decode` kind, `match` against the
generated string) or `distribution` (`top_k`, `cross_entropy`, `kl`, `class_probs`). Both
quantities are derived, never authored (§6), enter no digest, and `causalab explain` prints
them, so the requirement is a **value** that tests assert rather than a hidden optimization.

**How** a backend meets it is the backend's business — per-step captures, streamed keeps, a
teacher-forced replay, `logits_to_keep` (which accepts an index tensor and lowers to
`lm_head(hidden_states[:, idx, :])`), or a vocab-parallel reduction on a sharded backend. §8
already frames this as the backend's obligation ("derive minimal logit materialization
(`logits_to_keep`, vocab-parallel CE) from `save` + metric needs"); this issue extends that
clause to the decode and adds nothing normative about mechanism. The document and its digest
are identical either way.

For the reference backend, the natural first implementation needs no extra pass: read hooks
accumulate per decode step (writes stay prefill-only, so only capture hooks are live during the
decode), and the vocabulary projection is applied only at the positions that survive selection.
That is a backend note, not spec, and is swappable without touching a single document.

## Worked example — the probe, with no new section and no new metric kind

```json
"sites":     {"lm_head": {"component": "lm_head"}},
"positions": {"last_new": {"generated": {"max_new_tokens": 8}, "index": -1}},
"reads":     {"tail": {"site": "lm_head", "pos": "last_new",
                       "model": "patched", "input": "base"}},
"metrics":   {"greedy": {"kind": "top_k", "of": "tail", "k": 1}},
"save":      [{"value": "greedy", "model": "patched", "input": "base",
               "file_path": "probe.parquet"}]
```

One vocab slice materialized (tier 2), the read never saved (rule 11 counts a metric input as
a sink), text via an existing kind. Swap `last_new` for `said_answer` and the same document
reads *where the model said the expected answer*; point the read at `block_output` instead and
it harvests the residual stream there (tier 3).

## Implementation plan

1. **`causalab/protocol/schema.py`** — `PositionSpec` (`:365`) gains
   `generated: Leaf | None`; `_parse_position_spec` (`:846`) accepts it as a frame selector
   orthogonal to the anchors (`:854`, `:857`), requires exactly one anchor with it, and
   refuses `column`/`scope`/`relative_to` alongside it. Budget through `_wrapped` (sweeps
   free), int ≥ 1.
2. **`causalab/protocol/canonical.py`** — `_canon_position_spec` (`:231`) materializes the
   budget; positions already enter read closures, so digests separate budgets. Existing
   documents gain nothing: `tests/protocol/corpus_digests.json` must stay byte-identical.
3. **`causalab/protocol/validate.py`** — new rule: `generated` on a write's `pos` is a load
   error (prefill-only). Frame-awareness in `_provably_disjoint` (`:477`) and
   `_check_write_collisions` (`:557`). `train` + a `generated` position refuse (no gradients
   through a decode). A `generated` read at a non-`lm_head` site is **allowed** — tier 3
   covers it.
4. **`causalab/protocol/plan.py`** — `ForwardGroup` (`:64`) gains `decode_depth`
   (`max` budget over the group's generate positions) and `materialize`: the per-read
   position/site set plus whether a distribution is needed. `plan_point` (`:94`) derives both;
   `stop_after → None` for a generating group.
5. **`causalab/neural/pytorch_hooks/encoding.py`** — a `ContinuationFrame` beside
   `EncodedBatch` (`:55`): ids, per-row real widths (first-EOS), incremental char offsets.
   `resolve_position` (`:193`) dispatches on the frame; `variable_value` (`:123`) is reused
   verbatim for the value lookup.
6. **`causalab/neural/pytorch_hooks/executor.py`** — `_run_group` (`:237`): prefill with
   `use_cache=True` (the existing hooked forward), then the ids-only greedy loop with per-row
   EOS stop, tier-2 keeps inline, and one tier-3 replay when the materialization set asks for
   it. `_finalize_read` (`:352`) resolves continuation-frame positions. Decode steps run
   **un-hooked**; hooks come off when the prefill's `ExitStack` closes.
7. **`causalab/neural/pytorch_hooks/metrics.py`** — lift the one-position refusal
   (`_last_pos_logits`, `:189`) for continuation reads; per-step results; `step` and `matched`
   columns in the table (`outputs.py:125` `MetricTable`); a `domain` (`ids` |
   `distribution`) per kind; and one new ids-domain kind (`decode`) so text costs no
   projection.
8. **`causalab/protocol/backend.py`** — `generate` joins `CAPABILITIES` (`:34`) and is
   required (`:46`) when any position carries `generated`. Deliberately not `full_logits`:
   an argmax over the vocabulary lowers vocab-parallel and materializes nothing.
9. **`causalab/protocol/cli.py`** — `explain` prints the decode depth and the materialization
   set per group.
10. **`docs/intervention_protocol.md`** — §2.3: the frame selector, the four forms, the EOS /
    ragged rule, the `lm_head` off-by-one, the variable-match policy. §4: a generating group
    is one prefill + N decode steps + at most one replay. §5: the new rules. §8: the
    materialization tiers as the backend's obligation. §9: the `explain` addition. And
    **line 371's v1-scope sentence is rewritten** — prefill-only *interventions*, with greedy
    decode as an addressable frame.

## Tests

- **`tests/neural/pytorch_hooks/test_generate_frame.py`** (new):
  (a) un-intervened continuation matches `model.generate(do_sample=False, use_cache=True)`
  token-for-token on a two-row left-padded batch of unequal length, both tiny families;
  (b) **replay parity** — tier-3 activations and ids equal the decode's, so the teacher-forced
  pass is exact; and a write at `{"index": -1}` lands on the prompt's last token in the replay,
  not the continuation's (the sharp rule, as its own test);
  (c) intervened continuation differs from the un-intervened one under a large steer, and
  matches a prefill-guarded raw-hook oracle;
  (d) ragged widths: a row that emits EOS early contributes fewer positions, and `index: -1`
  is its last *real* token;
  (e) self-consistency: `top_k(k=1)` over an `all` read reproduces the decoded ids exactly.
- **Materialization is asserted as a plan, not as memory**: a `{"index": -1}` document's
  group plan names one step; an ids-only-metric document names no distribution; a `variable`
  document is marked replay. `tests/protocol/test_plan.py`.
- **`tests/protocol/test_schema.py`** — the four forms parse; `generated` without an anchor
  refuses; with two anchors refuses; with `column`/`scope`/`relative_to` refuses; budget `0`
  refuses; a swept budget parses.
- **`tests/protocol/test_validation_rules.py`** — one failing document each: a write with
  `generated`; `train` + `generated`. Plus the frame-disjointness case (a prompt-frame and a
  continuation-frame write position no longer collide) once writes are allowed there (they are
  not in v1 — assert the refusal instead).
- **`tests/neural/pytorch_hooks/test_positions_frame.py`** — variable found (spans cover the
  match), variable absent (width 0, `matched=false`, no crash), match spanning a
  sentencepiece merge.
- **Corpus**: `tests/protocols/10_probe_generate_im.json` (the worked example) through the
  real CLI on tiny-random; registered in `test_corpus.py`'s `CORPUS_SHAPE` with
  `requires == {"generate"}`; digest pinned. Existing nine digests byte-identical.

## Non-goals

- **Sampling.** Greedy only — no temperature, top-p, beams, penalties, stop strings. A sampled
  continuation is not a function of the document.
- **Decode-step writes.** Writes stay prefill-only; the intervention reaches the continuation
  through the first token's logits and what it left in the KV cache. `generated` on a write is
  a load error.
- **Stopping conditions beyond EOS + budget** (stop strings, `until_variable`).
- **`scope` / `relative_to` inside the continuation** — obvious follow-up, not v1.
- **Chat templates** — `prefix_lengths` stays 0; the same fidelity gap the golden tier already
  records.
- **`train` with a generated position**; multi-GPU; batching/perf work beyond the tiers above.

## Acceptance

- The worked example loads, validates, and runs on tiny-random through the CLI, producing a
  parquet with the greedy continuation, and its group plan names exactly one materialized step.
- A `variable`-anchored document reads `block_output` at the tokens where the model said the
  expected answer, is marked replay in the plan, and its replay activations equal the decode's.
- A row that never says the value yields width 0, `matched=false`, and a `null` metric — no
  refusal.
- Un-intervened continuations match `generate(do_sample=False)` token-for-token on both tiny
  families; intervened ones match a prefill-guarded raw-hook oracle.
- `causalab explain` reports decode depth, the materialization set, and
  `requires ['generate']`; a backend without the capability refuses with it named.
- `tests/protocol/corpus_digests.json` unchanged for the existing nine documents.
- Every new refusal (write with `generated`, `train` + `generated`, anchorless `generated`,
  `column`/`scope` with `generated`) has a one-failing-document test.

Depends on #20. Builds directly on #27's `all` spelling and its ragged read/save path, and on
#23's `<col>_variables` / column resolution. Related: #24 (component advertising) is
orthogonal; the metric `domain` field introduced here is the same mechanism I13 wants for
feature-space metrics, narrowed to one consumer.

**Comment by can-goodfire** [comment]
PR-2 is up as #44 (per-step metrics, the `ids` domain, and the `variable` anchor). Two things
in this issue's text are now out of date and should be read as superseded rather than built:

1. **The teacher-forced replay tier is unnecessary.** The design sketch above describes
   replaying the sequence to recover activations at generated positions, and lists "replay
   parity" in its acceptance. #40 removed the need: non-`lm_head` continuation reads are
   captured **per decode step** as the decode runs (`executor.py:288–299`), and an `lm_head`
   read is served from kept `ln_final` activations projected only at its addressed positions.
   There is nothing to replay, so there is nothing to prove parity against.

2. **The corpus document numbering.** The worked example is referred to as
   `10_probe_generate_im.json`; it landed as `tests/protocols/11_probe_generate_im.json` in
   #40, and #44 adds `12_probe_variable_im.json` for the scored/variable-anchored case.

Everything else in the spec above is what #40 and #44 implement.
